### DISCUSSION TITLE
StreamingResponse will not exit "with"-statement if user stops download

### DISCUSSION THREAD

**Discussion by benedikt-hess-km** [fastapi/fastapi#11151] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import os

import fastapi


app = fastapi.FastAPI()


@app.get("/")
def stream():
    return fastapi.responses.StreamingResponse(generator())


class Tester:
    def __enter__(self):
        print("Enter")

    def __exit__(self, exc_type, exc_val, exc_tb):
        print("Exit")


def generator():
    with Tester():
        while True:
            yield os.urandom(1024)
```

### Description

Hi everyone,

I am currently developing an app which zips files on demand and streams the resulting zip file. Unfortunately I am not able to handle a special case. If the user stops the download before it is finished the zipping process is not cleaned up, even when started within a "with" statement.

As my code is quite long i came up with the minimal example you can see above. I know this will run forever, all i want to show is, that the `__exit__` method is not reached if the user stops the download.

Steps to recreate:

* start app
* go to localhost:8000
* download will start
* cancel the download (you can even close the browser)
* check your output you can see "Enter" but no "Exit"

The same can be done with curl `curl localhost:8000 --output test.txt` and `CTRL+C` in your terminal.

As soon as the server is stopped "Exit" will be printed to the console!

Am I doing something wrong? Is there a better/another way of cleaning up if the stream is interrupted? Currently i am working with a watchdog which will kill the zipping process if no data was downloaded for more than X seconds, but I don't think this is the best way to do this.

Thank you for your help!

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

3.12.1

### Additional Context

App is served with uvicorn 0.27.1

**Comment by jgould22** [comment]
Your output is being buffered.

[Set the `PYTHONUNBUFFERED` ](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONUNBUFFERED)

or 

add a `flush=true`[ argument to your print statement](https://docs.python.org/3/library/functions.html#print).

**Reply by benedikt-hess-km** [reply]
Thank you for your answer! I changed  the line
```python
print("Exit")
```
to
```python
print("Exit", flush=True)
```
But when cancelling the download, "Exit" is not shown in the console. Once i stop the server, "Exit" is printed.
